presto-main/src/main/java/com/facebook/presto/operator/StageExecutionDescriptor.java

```diff
@@ -21,33 +21,67 @@
 import java.util.List;
 import java.util.Set;
 
+import static com.facebook.presto.operator.StageExecutionDescriptor.StageExecutionStrategy.DYNAMIC_LIFESPAN_SCHEDULE_GROUPED_EXECUTION;
+import static com.facebook.presto.operator.StageExecutionDescriptor.StageExecutionStrategy.FIXED_LIFESPAN_SCHEDULE_GROUPED_EXECUTION;
+import static com.facebook.presto.operator.StageExecutionDescriptor.StageExecutionStrategy.UNGROUPED_EXECUTION;
 import static com.google.common.base.Preconditions.checkArgument;
 import static java.util.Objects.requireNonNull;
 
 public class StageExecutionDescriptor
 {
+    private final StageExecutionStrategy stageExecutionStrategy;
     private final Set<PlanNodeId> groupedExecutionScanNodes;
 
-    private StageExecutionDescriptor(Set<PlanNodeId> groupedExecutionScanNodes)
+    private StageExecutionDescriptor(StageExecutionStrategy stageExecutionStrategy, Set<PlanNodeId> groupedExecutionScanNodes)
     {
-        this.groupedExecutionScanNodes = groupedExecutionScanNodes;
+        this.stageExecutionStrategy = requireNonNull(stageExecutionStrategy, "stageExecutionStrategy is null");
+        this.groupedExecutionScanNodes = requireNonNull(groupedExecutionScanNodes, "groupedExecutionScanNodes is null");
+
+        switch (stageExecutionStrategy) {
```

Comment:
Maybe extract this switch into a separate method and call it before assigning the fields.

e.g.:

```
private StageExecutionDescriptor(StageExecutionStrategy stageExecutionStrategy, Set<PlanNodeId> groupedExecutionScanNodes)
    {
        requireNonNull(stageExecutionStrategy, "stageExecutionStrategy is null");
        checkArgument(
                stageExecutionStrategy == UNGROUPED_EXECUTION && groupedExecutionScanNodes.isEmpty() ||
                        stageExecutionStrategy != UNGROUPED_EXECUTION && !groupedExecutionScanNodes.isEmpty(),
                "Unsupported stage execution parameters. stageExecutionStrategy: %s, groupedExecutionScanNodes: %s", 
                stageExecutionStrategy, 
                groupedExecutionScanNodes);
        this.stageExecutionStrategy = stageExecutionStrategy;
        this.groupedExecutionScanNodes = requireNonNull(groupedExecutionScanNodes, "groupedExecutionScanNodes is null");
    }
```

Reply:
@arhimondr : For such complicated check, explicitly using switch makes it easier to understand what get checked in each situation.

Moving the check is a good idea. Thanks!